src/Xrefcheck/Progress.hs

```diff
@@ -22,59 +28,127 @@ import Universum
+incProgressUnfixableErrors :: (Num a, Show a) => Progress a -> Progress a
+incProgressUnfixableErrors Progress{..} = Progress{ pErrorsUnfixable = pErrorsUnfixable + 1
+                                                  , ..
+                                                  }
+
+-- | Increase the number of fixable errors.
+incProgressFixableErrors :: (Num a, Show a) => Progress a -> Progress a
+incProgressFixableErrors Progress{..} = Progress{ pErrorsFixable = pErrorsFixable + 1
+                                                , ..
+                                                }
+
+-- | Decrease the number of fixable errors. This function indicates the situation where one of
+-- such errors had been successfully eliminated.
+decProgressFixableErrors :: (Num a, Show a) => Progress a -> Progress a
+decProgressFixableErrors Progress{..} = Progress{ pErrorsFixable = pErrorsFixable - 1
+                                                , ..
+                                                }
+
+fixableToUnfixable :: (Num a, Show a) => Progress a -> Progress a
+fixableToUnfixable Progress{..} = Progress{ pErrorsFixable = pErrorsFixable - 1
+                                          , pErrorsUnfixable = pErrorsUnfixable + 1
+                                          , ..
+                                          }
+
+setTaskTimestamp :: Time Second -> Time Second -> Progress a -> Progress a
+setTaskTimestamp ttc startTime Progress{..} = Progress{ pTaskTimestamp = Just (ttc, startTime)
+                                                      , ..
+                                                      }
+
+removeTaskTimestamp :: Progress a -> Progress a
+removeTaskTimestamp Progress{..} = Progress{ pTaskTimestamp = Nothing
+                                           , ..
+                                           }
+
+checkTaskTimestamp :: Time Second -> Progress a -> Progress a
+checkTaskTimestamp posixTime p@Progress{..} =
+  let mWithinGivenTime = do
+        (ttc, startTime) <- pTaskTimestamp
+        return $ ttc >= posixTime -:- startTime
+  in maybe p (\withinGivenTime ->
+    if withinGivenTime
+    then p
+    else removeTaskTimestamp p) mWithinGivenTime
 
 -- | Visualise progress bar.
-showProgress :: Text -> Int -> Color -> Progress Int -> Text
-showProgress name width col Progress{..} = mconcat
+showProgress :: Text -> Int -> Color -> Time Second -> Progress Int -> Text
+showProgress name width col posixTime Progress{..} = mconcat
   [ color col (name <> ": [")
   , toText bar
+  , timer
   , color col "]"
   , status
   ]
   where
     done = floor $ (pCurrent % pTotal) * fromIntegral width
-    errs = ceiling $ (pErrors % pTotal) * fromIntegral width
-    done' = max 0 $ done - errs
-    remained' = width - errs - done'
+    errsU = ceiling $ (pErrorsUnfixable % pTotal) * fromIntegral width
+    errsF = min (width - errsU) . ceiling $ (pErrorsFixable % pTotal) * fromIntegral width
+    done' = max 0 $ done - errsU - errsF
```

Comment:
While we're here, can you please rename `done'` to `successful`?

"done = done - errors" just looks wrong.